stream.c

```diff
@@ -0,0 +1,304 @@
+						encode->u.zeros.post_data_state =
+							COBS_ENCODE_STATE_ZEROS_FIRSTBYTE;
+					} else {
+						encode->u.zeros.post_data_state =
+							COBS_ENCODE_STATE_ZEROS_CODE;
+					}
+				}
+			} else {
+				encode->state = COBS_ENCODE_STATE_NOZEROS_DATA;
+				encode->u.nozeros = (struct cobs_encode_nozeros){
+					.total_length = num_processed,
+				};
+
+				const size_t total_length = num_processed;
+				if (total_length < 254) {
+					*output = encode->u.nozeros.total_length + 1;
+
+					encode->u.nozeros.data_left =
+						encode->u.nozeros.total_length;
+				} else {
+					*output = 0xFF;
+
+					encode->u.nozeros.data_left = 254;
+				}
+			}
+		} else if (encode->u.zeros.next_zero < 254) {
+			*output = encode->u.zeros.next_zero + 1;
+
+			encode->state = COBS_ENCODE_STATE_ZEROS_DATA;
+			encode->u.zeros.data_left = encode->u.zeros.next_zero;
+			encode->u.zeros.post_data_state = COBS_ENCODE_STATE_ZEROS_CODE;
+		} else {
+			*output = 0xFF;
+
+			encode->state = COBS_ENCODE_STATE_ZEROS_DATA;
+			encode->u.zeros.data_left = 254;
+			if (encode->u.zeros.next_zero == 254) {
+				encode->u.zeros.post_data_state = COBS_ENCODE_STATE_ZEROS_FIRSTBYTE;
+			} else {
+				encode->u.zeros.post_data_state = COBS_ENCODE_STATE_ZEROS_CODE;
+			}
+		}
+
+		break;
+	}
+
+	case COBS_ENCODE_STATE_ZEROS_DATA: {
+		size_t num_pulled = net_buf_pull_across_fragments(&encode->buf, output, 1);
+		__ASSERT_NO_MSG(num_pulled == 1);
+
+		encode->u.zeros.data_left -= 1;
+		encode->u.zeros.next_zero -= 1;
+
+		if (encode->u.nozeros.data_left == 0) {
+			encode->state = encode->u.zeros.post_data_state;
+		}
+		break;
+	}
+
+	case COBS_ENCODE_STATE_NOZEROS_CODE: {
+		if (encode->u.nozeros.total_length == 0) {
+			*output = 0x01;
+
+			encode->state = COBS_ENCODE_STATE_FINAL_ZERO;
+		} else if (encode->u.nozeros.total_length < 254) {
+			*output = encode->u.nozeros.total_length + 1;
+
+			encode->u.nozeros.data_left = encode->u.nozeros.total_length;
+			encode->state = COBS_ENCODE_STATE_NOZEROS_DATA;
+		} else {
+			*output = 0xFF;
+
+			encode->u.nozeros.data_left = 254;
+			encode->state = COBS_ENCODE_STATE_NOZEROS_DATA;
+		}
+		break;
+	}
+
+	case COBS_ENCODE_STATE_NOZEROS_DATA: {
+		size_t num_pulled = net_buf_pull_across_fragments(&encode->buf, output, 1);
+		__ASSERT_NO_MSG(num_pulled == 1);
+
+		encode->u.nozeros.data_left -= 1;
+		encode->u.nozeros.total_length -= 1;
+
+		if (encode->u.nozeros.total_length == 0) {
+			__ASSERT_NO_MSG(encode->u.nozeros.data_left == 0);
+			encode->state = COBS_ENCODE_STATE_FINAL_ZERO;
+		} else if (encode->u.nozeros.data_left == 0) {
+			encode->state = COBS_ENCODE_STATE_NOZEROS_CODE;
+		}
+		break;
+	}
+
+	case COBS_ENCODE_STATE_FINAL_ZERO:
+		*output = 0x00;
+		encode->state = COBS_ENCODE_STATE_FINISHED;
+		break;
+
+	case COBS_ENCODE_STATE_FINISHED:
+		return true;
+	}
+
+	return false;
+}
+
+size_t cobs_encode_stream(struct cobs_encode *encode, uint8_t *restrict output,
+			  size_t output_length)
+{
+	size_t i;
```

Comment:
Can the declaration be inside for loop ?

Reply:
no because we need to return it in line 303